### PR TITLE
fix(RHTAPBUGS-946): push floating tag when tagPrefix is set

### PR DESCRIPTION
**Pull request by mmalina**
When tagPrefix is set, in addition to the tagPrefix-timestamp tag, we also push the tagPrefix floating tag. This additional tag will also be created in Pyxis.

This depends on https://github.com/redhat-appstudio/release-service-utils/pull/86

### CODE REVIEW

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (mmalina) [inline tasks/create-pyxis-image/create-pyxis-image.yaml:62]: Change before merging.

File: tasks/push-snapshot/push-snapshot.yaml
Comment (mmalina) [inline tasks/push-snapshot/push-snapshot.yaml:152]: Question: Do we need this? Or should we still push using the timestamped tag only?

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (ralphbean) [inline tasks/create-pyxis-image/create-pyxis-image.yaml:102]: * If commonTag is specified, then the tags are 1) commonTag param and 2) tagPrefix from data file.
* If commonTag is _not_ specified, then the tags are 1) defaultTag from the data file.

It's hard for me to wrap my head around how these three inputs work together.

What's the usecase for the commonTag param? Do we use it anywhere? Ah - I see, it is a _result_ from push-snapshot task.

Perhaps push-snapshot's result should be _the list of tags, period_, and perhaps that list of tags should be used as a parameter here, eliminating the need for these conditionals in this task?

or, perhaps a task at the beginning of the pipeline should determine the list of tags, and that should be passed to both push-snapshot and to this task?

That refactoring doesn't need to happen here and now in this PR - but, I have a hard time being sure that this change is correct in review.

After talking it through, I fear it won't work. I think this change to this task will correctly create the tags in pyxis, but:

* The push-snapshot task never pushes the images to the registry with both the commonTag (the fixed tag) and the tagPrefix (the floating tag). I think it still only uses the commonTag. Confirm?
* The sign-image task never signs the tagPrefix (the floating tag). It only ever signs the commonTag (the fixed tag) like it does now. True?

It seems like `commonTag` should be refactored to `commonTags` or something like that, and be updated to be used across all three tasks.

Comment (ralphbean) [reply tasks/create-pyxis-image/create-pyxis-image.yaml:102]: I should have read on before I wrote all that. I see now that the push-snapshot task _is_ modified below.

I think the sign-image task may still need modification.

File: tasks/push-snapshot/push-snapshot.yaml
Comment (ralphbean) [reply tasks/push-snapshot/push-snapshot.yaml:152]: In the classic pipeline, only the equivalent of `${tagPrefix}-${timestamp}-source` is tagged. I don't see evidence of a source container tag for the floating tag in the registry.

Comment (mmalina) [reply tasks/push-snapshot/push-snapshot.yaml]: OK, thanks, removed.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (mmalina) [reply tasks/create-pyxis-image/create-pyxis-image.yaml:102]: I agree it's a bit cumbersome. We should definitely make it clearer. I'll create a follow up Jira if that's ok. But I think it should work now. And I wasn't sure about signing. So I added it now. Please check.

Comment (mmalina) [reply tasks/create-pyxis-image/create-pyxis-image.yaml:102]: To elaborate a bit more on how this works.

Before this change, the idea was: 

For rh releases, you want a common tag for all components in a release that includes some prefix, like 1.0.0.beta1, and a timestamp capturing the time when the release happened.

So you'd set a `tagPrefix` in your RPA. And then in the `push-snapshot` task there was logic, so that if `tagPrefix` was set, it will use `$tagPrefix-$timestamp` as the tag for all the components. And then this tag will be saved as the `commonTag` result and used for signing and creating the pyxis image.

After this change, I add the `tagPrefix` as an additional tag in all three places: When pushing the images, when signing them and when creating the images in Pyxis.

I think it was not very intuitive even before - the relationship between tagPrefix and commonTag (also called prefixedTag and now fixedTag) was not very clear.

Changing commonTag to commonTags and using it to share both tags at once to the other tasks would help. But I wonder if it could be improved further. If we push to both `tagPrefix` and `tagPrefix-timestamp`, is `tagPrefix` still a good name? How about calling it `version` or something like that? Also, why not merge it with `defaultTag`? Is there a reason why other releases shouldn't push tag-timestamp as well? I'm not a user of this, but it doesn't seem very intuitive to me that once you set tagPrefix, defaultTag is ignored. Or we could have defaultTag and enableTimestampTags which we would set for rh releases. Just some thoughts. If you think the options for RPA are good enough as they are for now, then we should probably go with the `commonTags` solution. Ping me and we can discuss more :)

Comment (ralphbean) [reply tasks/create-pyxis-image/create-pyxis-image.yaml:102]: All good thinking!

I really like the idea of ["ubiquitous language" from domain-driven design](https://martinfowler.com/bliki/UbiquitousLanguage.html). We should strive to align the terms in code and configuration with the terms that domain experts (product engineers and release engineers) use in real life. 

And, thinking about it now, the words we use in conversation about this are "floating tag" and "fixed tag".  1.0.beta is the "floating" tag (it is re-used on every push and so it "floats" forwards) but the 1.0.beta-<timestamp> tag is not re-used in practice and is unique for every push - it is "fixed" to one and only one image, in practice.

Something to consider for the followup JIRA.

File: tasks/push-snapshot/push-snapshot.yaml
Comment (ralphbean) [inline tasks/push-snapshot/push-snapshot.yaml:72]: Indentation differs here from two lines earlier, if that matters.

**Comment by ralphbean** [comment]
:+1: to merge and give it a test in stage on Monday.

File: tasks/push-snapshot/push-snapshot.yaml
Comment (mmalina) [reply tasks/push-snapshot/push-snapshot.yaml]: thanks, fixed

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (mmalina) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: I definitely like the idea of "ubiquitous language", thanks for the link. I will think about it for the follow up jira.

Meanwhile, I added a new commit to rename `commonTag` to `commonTags` and store both tags in it, without needing to read `tagPrefix` from the data file again in the consuming tasks (create-pyxis-image and rh-sign-image). Please take a look.

Comment (ralphbean) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: Ack. Just reviewed. Looks good!

**Review by ralphbean — APPROVED** [review]

**Pull request merged**
